Add GLDAS scripts & fix GLDAS job

KateFriedman-NOAA opened the pull request:
**Description**

This PR:

1. Adds the GLDAS scripts from the GLDAS repo (https://github.com/NOAA-EMC/GLDAS) into global-workflow (as described in issue #1014). The scripts are also
* Updated to conform to recent script cleanup and enhancements (e.g. add preamble).
* The scripts are removed from `.gitignore` and `link_workflow.sh`. 
* References to obsolete platforms (WCOSS_DELL_P3 and WCOSS_C) are removed.
* Updates for CFP were made to use existing CFP variables and remove machine checks.
* Updates were made to address linter/shellcheck warnings.
2. Fixes GLDAS job by updating fix file settings to use updated fix set provided by @HelinWei-NOAA . (Issue #622)

The following scripts are added to global-workflow:

* jobs/
  * JGDAS_ATMOS_GLDAS
* scripts/
  * exgdas_atmos_gldas.sh
* ush/
  * ~~gldas_archive.sh~~ (not bringing into global-workflow, see discussion below https://github.com/NOAA-EMC/global-workflow/issues/1014#issuecomment-1247045987)
  * gldas_forcing.sh
  * gldas_get_data.sh
  * gldas_liscrd.sh
  * ~~gldas_post.sh~~ (not bringing into global-workflow, see discussion below https://github.com/NOAA-EMC/global-workflow/issues/1014#issuecomment-1247045987)
  * gldas_process_data.sh

GLDAS scripts were first added to branch as-is and then updates were made to address the following:

- add preamble
- convert scripts from ksh to bash
- remove backticks and replace with substitutions
- remove unneeded announcement echos
- some general cleanup
- resolve linter/shellcheck warnings

Fixes #622
Resolves #1014

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

- [x] Clone and build tests on Orion
- [x] C192C96L127 cold-started cycled test on Orion (21.5 cycles = 16.5 cycle spinup for GLDAS + 5 cycles after spinup with 2 00z cycles with GLDAS invoked)
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

KateFriedman-NOAA committed:
Remove GLDAS scripts from .gitignore

- Remove GLDAS JJOB script from ignore file.
- Remove GLDAS exscript from ignore file.
- Remove GLDAS ush scripts from ignore file.

Refs #1014

KateFriedman-NOAA committed:
Remove GLDAS script symlinks from link_workflow.sh

- Remove symlink creation for GLDAS scripts in the workflow link script.
- Removal includes JJOB, ex-script, and ush script symlinks.

Refs #1014

KateFriedman-NOAA committed:
Add GLDAS scripts to global-workflow

- Initial copy of GLDAS JJOB, exscript, and ush scripts from GLDAS
repo into global-workflow.

Refs #1014

KateFriedman-NOAA committed:
Updates to GLDAS scripts

- Convert from ksh to bash.
- Convert backticks to substitutions.
- Add preamble.
- Remove announcements and general cleanup.

Refs #1014

KateFriedman-NOAA committed:
Cleanup GLDAS script doc blocks more

Refs #1014

KateFriedman-NOAA committed:
Address linter warnings and remove obsolete platforms

- Remove references to WCOSS_DELL_P3 and WCOSS_C in GLDAS scripts.
- Update $HOMEgfs variable in GLDAS preable source lines to address
warnings from linter regarding:
1) HOMEgfs is referenced but not assigned.
2) Prefer putting braces around variable references even when not strictly required.

Refs #1014

KateFriedman-NOAA committed:
Address linter warnings in JGDAS_ATMOS_GLDAS

Refs #1014

KateFriedman-NOAA committed:
Address linter warning for KEEPDATA and assign gldas_spinup_hours

Refs #1014

KateFriedman-NOAA committed:
Address linter warning for SC1091 in JGDAS_ATMOS_GLDAS

Refs #1014

KateFriedman-NOAA committed:
Update linter directive for machine env line in JGDAS_ATMOS_GLDAS

Refs #1014

KateFriedman-NOAA committed:
Make machine for env file uppercase

Refs #1014

KateFriedman-NOAA committed:
Address linter warnings in scripts/exgdas_atmos_gldas.sh

- Address SC2016 (double instead of single quote for ERRSCRIPT
and SC2250 (curly braces).

Refs #1014

KateFriedman-NOAA committed:
Resolve SC2250 (curly brace) shellcheck warnings in GLDAS ush scripts

Refs #1014

KateFriedman-NOAA committed:
Address SC2086 shellcheck warnings in GLDAS ush scripts

Refs #1014

KateFriedman-NOAA committed:
Correct env sourcing in jobs/JGDAS_ATMOS_GLDAS

Refs #1014

KateFriedman-NOAA committed:
Address SC2086 warnings in scripts/exgdas_atmos_gldas.sh

Refs #1014

KateFriedman-NOAA committed:
Address SC2086 warnings in ush/gldas_get_data.sh

Refs #1014

KateFriedman-NOAA committed:
Address a number of linter warnings in GLDAS scripts

Refs #1014

KateFriedman-NOAA committed:
Address some linter warnings in GLDAS scripts

- Mostly address "referenced but not assigned" warnings

Refs #1014

KateFriedman-NOAA committed:
Address linter warnings in ush/gldas_archive.sh and ush/gldas_post.sh

Refs #1014

KateFriedman-NOAA committed:
A couple syntax corrections in ush/gldas_forcing.sh

Refs #1014

KateFriedman-NOAA committed:
Address a couple linter warnings

Refs #1014

KateFriedman-NOAA committed:
Removing gldas_archive.sh and gldas_post.sh

After discussion with GLDAS code manager and confirming what GLDAS
scripts are actually used at runtime, it was decided to not bring
the ush/gldas_archive.sh and ush/gldas_post.sh scripts from GLDAS
into global-workflow.

Deleting them from this branch.

Refs #1014

KateFriedman-NOAA committed:
Address linter warnings in jobs/JGDAS_ATMOS_GLDAS and ush/gldas_forcing.sh

Refs #1014

KateFriedman-NOAA committed:
Convert GLDAS script to replace machine checks with CFP variables

- Remove checks for machine in ush/gldas_get_data.sh and replace with
existing CFP variables in system (USE_CFP and CFP_MP).
- Update Hera and Orion env files to set USE_CFP and CFP_MP as needed.
- Similar env variable settings will be added into WCOSS2.env in that branch.

Refs #1014

KateFriedman-NOAA committed:
Address double quote linter warning in ush/gldas_get_data.sh

Refs #1014

KateFriedman-NOAA committed:
Remove exports of local variables in scripts/exgdas_atmos_gldas.sh

Refs #1014

KateFriedman-NOAA committed:
Removing exports for local variables in ush/gldas_liscrd.sh

Refs #1014

KateFriedman-NOAA committed:
Remove unnecessary variable setting in ush/gldas_liscrd.sh

Refs #1014

KateFriedman-NOAA committed:
GLDAS fix file updates for FRAC_GRID

- Updates to set GLDAS fix files based on FRAC_GRID setting.
- Update scripts/exgdas_atmos_gldas.sh to set FIXgldas subfolder
based on value of FRAC_GRID: use "frac_grid" subfolder when .true.
or use "nonfrac_grid" subfolder when .false.
- Move FRAC_GRID variable from config.fcst to config.base since it will
now be used in more than just the forecast job.
- Also correct FIXgldas default in scripts/exgdas_atmos_gldas.sh to
add missing "/fix_gldas" path.

Refs #1014

KateFriedman-NOAA committed:
Address linter warnings and use iter as argument to gldas_process_data.sh

- Update GLDAS scripts and config.fcst to address linter/shellcheck warnings.
- Update gldas_get_data.sh and gldas_process_data.sh to pass $iter variable
as argument in gldas_process_data.sh and give to preamble.

Refs #1014

KateFriedman-NOAA committed:
Address linter warnings in config.fcst for SC2108

Refs #1014

KateFriedman-NOAA committed:
Add shellcheck disable for SC1091 in JGDAS_ATMOS_GLDAS

Refs #1014

KateFriedman-NOAA committed:
Also disable SC1090 in JGDAS_ATMOS_GLDAS

Refs #1014

KateFriedman-NOAA commented:
@aerorahul @WalterKolczynski-NOAA This PR is ready for your review. Since initiating it I have:

1. Addressed and resolved all linter/shellcheck warnings.
2. Adjusted how the GLDAS scripts use CFP but incorporating existing CFP variables (`USE_CFP` and `CFP_MP`) to remove machine checks.
3. Fixed GLDAS job (issue #622) by testing updated GLDAS fix files provided by @HelinWei-NOAA and updating GLDAS scripts to use them based on the value of the `FRAC_GRID` variable.

This PR will now resolve issues #622 and #1014.

WalterKolczynski-NOAA commented on jobs/JGDAS_ATMOS_GLDAS:
```diff
@@ -0,0 +1,136 @@
+
+export DATA=${DATA:-${DATAROOT}/${jobid:?}}
+mkdir -p "${DATA}"
+cd "${DATA}" || exit
```
This should be a non-zero exit since it is an error.

KateFriedman-NOAA replied:
Added "1" with the exit:

```
cd "${DATA}" || exit 1
```

WalterKolczynski-NOAA commented on jobs/JGDAS_ATMOS_GLDAS:
```diff
@@ -0,0 +1,136 @@
+##########################################
+# Remove the Temporary working directory
+##########################################
+cd "${DATAROOT}" || exit
```
Theoretically so should this one, but I don't know if we want to job to "fail" it it just can't delete `${DATA}`.

KateFriedman-NOAA replied:
Added "1" with the exit:

```
cd "${DATA}" || exit 1
```

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -252,9 +252,8 @@ export DO_OCN_SPPT=${DO_OCN_SPPT:-"NO"}
-export FRAC_GRID=".true."
 export cplmode="nems_frac"
-if [ $FRAC_GRID = ".false." ]; then
+if [[ "${FRAC_GRID:-".true."}" = ".false." ]]; then
```
Does this work as intended? Seems like there might be a quotation mark issue on the left side.

KateFriedman-NOAA replied:
I reran the gdasfcst job and it appears to work. Here are the lines from the log where line 256 in config.fcst is invoked:

```
++ config.fcst[255]: export cplmode=nems_frac
++ config.fcst[255]: cplmode=nems_frac
++ config.fcst[256]: [[ .true. = \.\f\a\l\s\e\. ]]
++ config.fcst[259]: export psm_bc=1
```

WalterKolczynski-NOAA commented on scripts/exgdas_atmos_gldas.sh:
```diff
@@ -0,0 +1,337 @@
+export OMP_NUM_THREADS=1
+export pgm=gdas2gldas
+# shellcheck disable=SC1091
+. prep_step
```
We should probably look at whether this actually needs to be sourced, or just run normally. And does it actually have to be called before each step? Can be later though.

KateFriedman-NOAA replied:
Will check in later work.

WalterKolczynski-NOAA commented on ush/gldas_forcing.sh:
```diff
@@ -0,0 +1,120 @@
+sdate=${bdate}
+edate=$(sh "${FINDDATE:?}" "${edate}" d-1)
+while [[ "${sdate}" -lt "${edate}" ]] ; do
+#-------------------------------
```
Indent block

KateFriedman-NOAA replied:
Indented.

WalterKolczynski-NOAA commented on ush/gldas_get_data.sh:
```diff
@@ -0,0 +1,76 @@
+  [[ ! -d ${fpath}/gdas.${ymd} ]] && mkdir -p "${fpath}/gdas.${ymd}"
+  [[ ! -d ${gpath}/gdas.${ymd} ]] && mkdir -p "${gpath}/gdas.${ymd}"
+
+f=1
```
This is still part of the while block but lost its indentation.

KateFriedman-NOAA replied:
Indented.

KateFriedman-NOAA committed:
Indent while loops in GLDAS scripts

- Add missing indentation into ush/gldas_forcing.sh and ush/gldas_get_data.sh.

Refs #1014

aerorahul commented on scripts/exgdas_atmos_gldas.sh:
```diff
@@ -0,0 +1,337 @@
+#################################
+# Set up UTILITIES
+#################################
+export FINDDATE=${FINDDATE:-/apps/ops/prod/nco/core/prod_util.v2.0.13/ush/finddate.sh}
```
Why are wcoss2 paths hard-wired here and below?
I guess we can handle that in another PR.
Would you add a comment 
```
#TODO - remove hard-wired WCOSS2 default paths
```

KateFriedman-NOAA replied:
Good catch. Those defaults should use module-provided paths to those utilities. I will correct in a subsequent PR. Thanks!

aerorahul reviewed (approved):
Mostly looks good.
This PR basically brings in scripts from the GLDAS repo into global-workflow, and applies the linter recommended standards.

KateFriedman-NOAA committed:
Provide exit code values in JGDAS_ATMOS_GLDAS

- Provide a value of "1" for new exit statements in GLDAS JJOB script.

Refs #1014
